### PR TITLE
Skip diagnostic passes for deserialized functions.

### PR DESCRIPTION
**Pull request by atrick**
Establish control over the pass pipeline that runs on serialized functions. Certain passes should definitely not be rerun. Most optimization passes should not need to be rerun. The assert configuration makes this tricky though. And PredictableMemOps either has some pass order problems or misses some things the first time, so we rerun it for the best performance--this is really unfortunate for compile time though.

### CODE REVIEW

**Comment by atrick** [comment]
@swift-ci Please test compiler performance.

**Comment by atrick** [comment]
@swift-ci benchmark.

**Comment by atrick** [comment]
@swift-ci Please test compiler performance.

**Comment by atrick** [comment]
@swift-ci Please test compiler performance

File: lib/SILOptimizer/PassManager/PassManager.cpp [hunk 1]
```diff
@@ -298,6 +298,10 @@ void SILPassManager::runPassOnFunction(unsigned TransIdx, SILFunction *F) {
 
   assert(analysesUnlocked() && "Expected all analyses to be unlocked!");
 
+  // Serialization requires all diagnostic passes. Don't rerun them.
+  //!!!if (isMandatoryPipeline && F->wasDeserialized())
+  //!!!  return;
+
```
Comment (eeckstein) [inline lib/SILOptimizer/PassManager/PassManager.cpp]: Did you forget to remove these lines?

**Comment by atrick** [comment]
TODO
- Rename wasDeserialized to isBornCanonical
- Cleanup isSIB comments. AFAICT there's currently no indication of the SIL stage in the serialization format. So we'll continue to assume raw SIL for now.

**Comment by jrose-apple** [comment]
Oops, I thought we did record -emit-sibgen vs. -emit-sib, but apparently not. No reason why we couldn't, though.

**Comment by atrick** [comment]
@swift-ci Please smoke test compiler performance

**Comment by atrick** [comment]
@swift-ci benchmark

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@jrose-apple, nothing much to see here, but I did rename the `wasDeserializedCanonical` flag and fixed up the comments/commit log.

**Comment by atrick** [comment]
@swift-ci test linux platform

**Comment by atrick** [comment]
@swift-ci benchmark

**Comment by atrick** [comment]
@swift-ci smoke test compiler performance

**Comment by atrick** [comment]
@swift-ci test OS X platform

**Comment by atrick** [comment]
@swift-ci test OS X platform

**Comment by atrick** [comment]
@swift-ci smoke test compiler performance

**Comment by atrick** [comment]
@swift-ci test OS X platform

**Comment by atrick** [comment]
@swift-ci test OS X platform

**Comment by atrick** [comment]
@shahmishal, that's five failed attempts to test on OS X. Not sure what the problem is.

**Comment by atrick** [comment]
WORKSPACE: /Users/buildnode/jenkins/workspace/swift-PR-compiler-performance-smoke-test-macOS
No such file or directory.

**Comment by atrick** [comment]
@swift-ci smoke test compiler performance

**Comment by atrick** [comment]
We won't see much benefit from this yet--linking less eagerly and removing assert configuration checks later would have regressed some benchmarks-- but it gives us the pipeline control that we need and lets me enable exclusivity verification. See https://bugs.swift.org/browse/SR-6961.

**Comment by atrick** [comment]
The benchmark report is only noise. As-is this doesn't have any noticeable affect on performance (which is ideal), and doesn't significantly affect compile time or code size. Those benefits will come when we stop eagerly linking during mandatory inlining and disable rerunning of PredictableMemOps and DiagnosticConstantProp.

**Pull request merged**